src/charm.py

```diff
@@ -401,15 +372,62 @@ def _add_ca_certificate_to_pki_secrets_engine(self) -> None:
 
     def _sync_vault_pki(self) -> None:
         """Goes through all the vault-pki relations and sends necessary TLS certificate."""
-        for relation in self.model.relations[PKI_RELATION_NAME]:
-            outstanding_requests = self.vault_pki.get_outstanding_certificate_requests(
-                relation_id=relation.id
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki request")
+            return
+        outstanding_pki_requests = self.vault_pki.get_outstanding_certificate_requests()
+        for pki_request in outstanding_pki_requests:
+            self._generate_pki_certificate_for_requirer(
+                csr=pki_request.csr,
+                relation_id=pki_request.relation_id,
             )
-            for request in outstanding_requests:
-                self._generate_pki_certificate_for_requirer(
-                    csr=request.csr,
-                    relation_id=relation.id,
-                )
+
+    def _sync_vault_kv(self) -> None:
+        """Goes through all the vault-kv relations and sends necessary KV information."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-kv request")
+            return
+        outstanding_kv_requests = self.vault_kv.get_outstanding_kv_requests()
+        for kv_request in outstanding_kv_requests:
+            relation = self.model.get_relation(relation_name = KV_RELATION_NAME, relation_id=kv_request.relation_id)
+            if not relation:
+                logger.warning("Relation not found for relation id %s", kv_request.relation_id)
+                continue
+            self._generate_kv_for_requirer(
+                relation=relation,
+                app_name=kv_request.app_name,
+                unit_name=kv_request.unit_name,
+                mount_suffix=kv_request.mount_suffix,
+                egress_subnet=kv_request.egress_subnet,
+                nonce=kv_request.nonce,
+            )
+
+    def _generate_kv_for_requirer(
+            self,
+            relation: Relation,
+            app_name: str,
+            unit_name: str,
+            mount_suffix: str,
+            egress_subnet: str,
+            nonce: str,
+        ):
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-kv request")
+            return
+        ca_certificate = self.tls.pull_tls_file_from_workload(File.CA)
+        if not ca_certificate:
+            logger.debug("Vault CA certificate not available")
+            return
+        vault = self._get_initialized_vault_client()
+        if not vault:
+            logger.debug("Failed to get initialized Vault")
+            return
+        vault.enable_approle_auth_method()
+        mount = "charm-" + app_name + "-" + mount_suffix
```

Comment:
Pretty minor again, but this would be more efficient by not creating 3 different strings.
```suggestion
        mount = f"charm-{app_name}-{mount_suffix}"
```